EnvironmentSimulator/Modules/CommonMini/UDP.cpp

```diff
@@ -12,6 +12,12 @@
 
 #include <stdio.h>
 
+#ifdef _WIN32
+	#include <winsock.h>
```

Comment:
Would it work just as good with winsock2.h?

Now there seems to be clash between winsock2.h, included by UDP.hpp and winsock.h included by UDP.cpp.

My suggestion is to change:

```
#ifdef _WIN32
	#include <winsock.h>
#else
	#include <sys/time.h>
#endif
```
into:
```
#ifndef _WIN32
	#include <sys/time.h>
#endif
```

Reply:
it's a good idea

Reply:
Great that you fixed the UDP module, thanks.

Now, I tried to build the project. But it fails on the last step for esmini.js.

Could you maybe provide:
1. What system you run on, Windows or Linux?
2. If Windows, what environment: git bash, PowerShell or simple CommandPrompt?
3. A simple list of the steps you perform

That would be very appreciated!

Reply:
sorry, i use macos

Reply:
i'm sorry, i build use window is fails,but i use wsl2 ubuntu18.04 build succ.

system info:
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.6 LTS
Release:        18.04
Codename:       bionic

gcc:
gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

g++:
g++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

emcc(https://emscripten.org/docs/getting_started/downloads.html):
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.28 (f11d6196dd4e8748a726f19895c859b40ff6a4f3)
Copyright (C) 2014 the Emscripten authors (see AUTHORS.txt)
This is free and open source software under the MIT license.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Reply:
build with liunx
1、install emscripten（https://emscripten.org/docs/getting_started/downloads.html）
```sh
git clone https://github.com/emscripten-core/emsdk.git
cd emsdk
# Fetch the latest version of the emsdk (not needed the first time you clone)
git pull

# Download and install the latest SDK tools.
./emsdk install latest

# Make the "latest" SDK "active" for the current user. (writes .emscripten file)
./emsdk activate latest

# Activate PATH and other environment variables in the current terminal
source ./emsdk_env.sh
```
2、install other system tools and libraries 
```
sudo apt install cmake  bzip2
```
3、build js 
```sh
cd esmini/EnvironmentSimulator/Libraries/esminiJS
./build.sh
```
4、check esmini.js
```
ll build/esmini.js
```

Reply:
Thanks for the guide! I'm trying it on the three platforms. 

Question: Do you really need the "'" in the CMakeLists.txt?

I would like to change line 82:
`-s 'EXPORT_NAME=\"esmini\"' \`
to:
`-s EXPORT_NAME=\"esmini\" \`

and line 83:
`-s EXPORTED_RUNTIME_METHODS='[\"cwrap\"]' \`
to:
`-s EXPORTED_RUNTIME_METHODS=[\"cwrap\"] \`

With this change it might work on Windows (cmd) as well. Can you check if this change is OK on your side as well?

I tried on macOS Catalina, and it seems to work fine with those changes.

Reply:
For EXPORTED_RUNTIME_METHODS, I use
-sEXPORTED_RUNTIME_METHODS=[cwrap]
and it works on Windows and Linux.

For EXPORT_NAME, I guess it should be fine to use:
-sEXPORT_NAME=esmini
for both Windows and Linux.

Reply:
Ok, so you mean you don't even need the escaped quotes, \"?
So that line 82:
`-s 'EXPORT_NAME=\"esmini\"' \`
to:
`-s EXPORT_NAME=esmini \`
and line 83:
`-s EXPORTED_RUNTIME_METHODS='[\"cwrap\"]' \`
to:
`-s EXPORTED_RUNTIME_METHODS=[cwrap] \`
?

Anyway I made the first suggested update to the CMakeFiles.txt. And also removed the generated esmini.js from example folder. Although it's nice to have a pre-built working example, it's too big I think, and also it would need to be updated now and then to stay in synch with esmini development.

I have now checked all three and they all work! 🙂

I updated the readme.md with my notes. Please check and make corrections if you find errors or other issues.

Let me know if you're OK with the changes. Then I think we're ready to merge the pull request.

Last question: I guess graphics 3D view is not expected in this example. So that might be a goal to achieve in future?

Reply:
thinks, i build with windows succ, graphics 3D view i do in 2023 maybe。
readme.md is weel good, i learn with you.

Anyway, bests wish.